testing/ngr.py

```diff
@@ -187,6 +190,55 @@ def test(self) -> None:
         run_command(f'dotnet test --framework={self.framework} --collect:"XPlat Code Coverage"')
 
 
+class JavaRunner(RunnerBase):
+    """
+    Java test suite runner.
+
+    - Knows how to invoke either Gradle or Maven.
+    """
+
+    def __post_init__(self) -> None:
+        self.has_pom_xml = None
+        self.has_gradle_files = None
+
+    def peek(self) -> None:
+        self.has_pom_xml = mp(self.path, "pom.xml")
+        self.has_gradle_files = mp(self.path, "*.gradle")
+
+        if self.has_pom_xml or self.has_gradle_files:
+            self.type = ItemType.JAVA
+
+    def info(self) -> None:
+        """
+        Display environment information.
+        """
+        run_command("java -version")
+
+    def install(self) -> None:
+        """
+        Install dependencies.
+        """
+        if self.has_pom_xml:
+            run_command("mvn install")
+        elif self.has_gradle_files:
+            run_command("./gradlew install")
+        else:
+            raise NotImplementedError("Unable to invoke target: install")
+
+    def test(self) -> None:
+        """
+        Invoke software tests.
+        """
+        if self.has_makefile:
+            run_command("make test")
+        elif self.has_pom_xml:
+            run_command(f'mvn test')
+        elif self.has_gradle_files:
+            run_command("./gradlew check")
```

Comment:
to run gradle tests, one would issue `./gradlew test`. Or what do I miss?

Reply:
Thanks for spotting. I may have picked this up somewhere else.

Reply:
Maybe `test` would invoke the tests, but `check` invokes both the linters, **and** the software tests?

Reply:
> The reason _check_ is important is because it can aggregate one or more verification tasks together. This allows us to incorporate multiple tasks into a single step rather than having to execute them one by one.
>
> The Gradle ecosystem contains a number of plugins, such as the [checkstyle](https://docs.gradle.org/current/userguide/checkstyle_plugin.html) plugin, that provide additional functionality to the _check_ task.

-- https://www.baeldung.com/gradle-test-vs-check

Reply:
Ah ok I see, normally I'd prefer to explicit name tasks one want to run, but as a generic task command this looks feasible.

Reply:
Keep in mind that `./gradle check` doesn't work correctly for CrateDB (talking about < 5.5 versions, that use gradle`:
```
~/crate/crate-5.4 (5.4 ✘)⚑✭ ᐅ ./gradlew clean check                                                                                                                                                                                                                 (5.4|✭1)
Starting a Gradle Daemon (subsequent builds will be faster)
....................................................................................................................................................................................................................................................................
> Task :libs:shared:compileJava
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
.........................................

> Task :libs:guice:compileJava
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :libs:sql-parser:compileJava
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

> Task :libs:shared:forbiddenApisTest FAILED

> Task :server:compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libs:shared:forbiddenApisTest'.
> No signatures were added to task; use properties 'signatures', 'bundledSignatures', 'signaturesURLs', and/or 'signaturesFiles' to define those!

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 51s
93 actionable tasks: 75 executed, 18 up-to-date

```

Reply:
Thanks. `ngr` will need to handle CrateDB differently then. I will consider this for another iteration. Up until now, it did not need to build CrateDB yet, only auxiliary artefacts.

Reply:
yep, it was just FYI, if you ever stumble upon this.